azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java

```diff
@@ -0,0 +1,185 @@
+      if (execDir != null) {
+        cleanup(execDir);
+      }
+      Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Prepare the project directory.
+   *
+   * @param pv ProjectVersion object
+   * @throws ProjectManagerException
+   * @throws IOException
+   */
+  @VisibleForTesting
+  void setupProject(final ProjectVersion pv)
+      throws ProjectManagerException, IOException {
+    final int projectId = pv.getProjectId();
+    final int version = pv.getVersion();
+
+    final String projectDir = String.valueOf(projectId) + "." + String.valueOf(version);
+    if (pv.getInstalledDir() == null) {
+      pv.setInstalledDir(new File(projectsDir, projectDir));
+    }
+
+    // If directory exists. Assume its prepared and skip.
+    if (pv.getInstalledDir().exists()) {
+      log.info("Project already cached. Skipping download. " + pv);
+      return;
+    }
+
+    log.info("Preparing Project: " + pv);
+
+    File tempDir = new File(projectsDir, "_temp." + projectDir + "." + System.currentTimeMillis());
+
+    // TODO Why mkdirs? This path should be already set up.
+    tempDir.mkdirs();
+
+    ProjectFileHandler projectFileHandler = null;
+    try {
+      projectFileHandler = requireNonNull(projectLoader.getUploadedFile(projectId, version));
+      checkState("zip".equals(projectFileHandler.getFileType()));
+
+      log.info("Downloading zip file.");
+      final File zipFile = requireNonNull(projectFileHandler.getLocalFile());
+      final ZipFile zip = new ZipFile(zipFile);
+      Utils.unzip(zip, tempDir);
+
+      Files.move(tempDir.toPath(), pv.getInstalledDir().toPath(), StandardCopyOption.ATOMIC_MOVE);
+
+      log.warn(String.format("Project Preparation complete. [%s]", pv));
+    } finally {
+
+      if (projectFileHandler != null) {
+        projectFileHandler.deleteLocalFile();
+      }
+
+      // Clean up: Remove tempDir if exists
+      if (tempDir.exists()) {
+        FileUtils.deleteDirectory(tempDir);
+      }
+    }
+  }
+
+  private void copyCreateHardlinkDirectory(File projectDir, File execDir) throws IOException {
+    FileIOUtils.createDeepHardlink(projectDir, execDir);
+  }
+
+  private File createExecDir(ExecutableFlow flow) {
+    final int execId = flow.getExecutionId();
+    File execDir = new File(executionsDir, String.valueOf(execId));
+    flow.setExecutionPath(execDir.getPath());
+
+    // TODO Why mkdirs? This path should be already set up.
+    execDir.mkdirs();
+    return execDir;
+  }
+
+  private ProjectVersion getProjectVersion(ExecutableFlow flow) {
+    // We're setting up the installed projects. First time, it may take a while
+    // to set up.
+    final ProjectVersion projectVersion;
+    synchronized (installedProjects) {
+      projectVersion = installedProjects.computeIfAbsent(new Pair<>(flow.getProjectId(), flow.getVersion()),
+          k -> new ProjectVersion(flow.getProjectId(), flow.getVersion()));
+    }
+    return projectVersion;
+  }
+
+  private void cleanup(File execDir) {
```

Comment:
the cleanup method is basically equivalent of FileUtils.deleteDirectory(execDir), maybe we can remove it?